client/service/src/chain_ops.rs

```diff
@@ -100,10 +273,21 @@ impl<
 			}
 		}
 
-		let mut io_reader_input = IoReader(input);
-		let mut count = None::<u64>;
-		let mut read_block_count = 0;
 		let mut link = WaitLink::new();
+		let block_stream_res: Result<BlockStream<_, Self::Block>, String> = BlockStream::new(input, binary);
+
+		let mut block_stream;
+		match block_stream_res {
+			Ok(bs) => {block_stream = bs},
+			Err(e) => {
+				// We've encountered an error while creating the block stream 
+				// so we can just return a future that returns an error.
+				let err_import = future::poll_fn(move |_cx| {
+					return std::task::Poll::Ready(Err(Error::Other(e.clone())))
+				});
+				return Box::pin(err_import)
+			}
+		}
```

Comment:
Would be keen to hear reviewers' thoughts on this particular piece of code. It seems a little bit hacky and I'm wondering if there's a better way to write this :)